### PR TITLE
feat: adds slack-user service 🚚

### PR DESCRIPTION
**Pull request by omar-elbaz**
## Description ✏️

Updates Slack User Service with two new functions. GetAllUsers and GetUserProfile

Related to #2 

- Implement two new functions in the slack-user.service file which call the slack api to:
- Get All Users in the ColorStack Family Workspace
- Fetch each user's individual User Profile

These will be needed in order to migrate birthdates from Slack to the ColorStack Member Profile Database, and will also be beneficial in the future when accessing Slack users.

Adjusted to implement our oyster RateLimiter to handle Slack Api's restrictions

## Type of Change 🐞

- [x] Feature - A non-breaking change which adds functionality.
- [ ] Fix - A non-breaking change which fixes an issue.
- [ ] Refactor - A change that neither fixes a bug nor adds a feature.
- [ ] Documentation - A change only to in-code or markdown documentation.
- [ ] Tests - A change that adds missing unit/integration tests.
- [ ] Chore - A change that is likely none of the above.

## Checklist ✅

- [x] I have done a self-review of my code.
- [ ] I have manually tested my code (if applicable).
- [ ] I have added/updated any relevant documentation (if applicable).

### CODE REVIEW

**Comment by ramiAbdou** [comment]
@omar-elbaz please make sure to check all the relevant boxes in the checklist of the PR! It's always great practice to self-review your own PR so that first one will be mandatory! ✅ I'll update the documentation to make that a bit clearer.

**Comment by omar-elbaz** [comment]
noted

**Review by ramiAbdou — CHANGES_REQUESTED** [review]
Great stuff, Omar! 🚀

I only have one concern with the approach. Currently, the plan is:
1. Get all the users in the Slack workspace.
2. Fetch each individual Slack profile.
3. Update the database based on the `slackId` of a student (I'm assuming).

The issue with that approach is there are a bunch of users in the Slack workspace that aren't associated with a member in our database (for various reasons), so we'd actually be hitting the Slack API more than we need to. Instead, we can:
1. Fetch all the `slackId`s from our database. This is one SQL query that is super quick and much more efficient than pulling ID's using the Slack API.
2. Fetch each individual Slack profile.
3. Update the database based on the `slackId` of a student.

The latter approach will ensure that we're not doing more work than we need to.

Let me know if you have any questions on that!

File: packages/core/src/modules/slack/services/slack-user.service.ts [hunk 1]
```diff
@@ -35,3 +43,47 @@ export async function updateSlackEmail(id: string, email: string) {
...
+
+      if (!users) {
+        break;
+      }
+
+      allUsers.push(...users);
+
+      // Update cursor to the next cursor value, if any.
+      cursor = response.response_metadata?.next_cursor;
+    } while (cursor);
```
[the remark this answers is not in the document]
Comment (omar-elbaz) [reply packages/core/src/modules/slack/services/slack-user.service.ts:68]: No strong preference, this looks good. I will make the change.

File: packages/core/src/modules/slack/services/slack-user.service.ts [hunk 2]
```diff
@@ -1,5 +1,13 @@
 import { slack } from '@/modules/slack/instances';
 import { ENV } from '@/shared/env';
+import { Member } from '@slack/web-api/dist/response/UsersListResponse';
+import { Profile } from '@slack/web-api/dist/response/UsersProfileGetResponse';
```
[the remark this answers is not in the document]
Comment (omar-elbaz) [reply packages/core/src/modules/slack/services/slack-user.service.ts:4]: I don't think that works. 
image

**Comment by ramiAbdou** [comment]
@omar-elbaz I'm going to close your PR and feel free to re-open once it's ready to be re-reviewed! I'm happy to chat through anything to drive any additional clarity as well.

**Pull request closed without merging**